### PR TITLE
INSP: Assign to immutable inspection

### PR DESCRIPTION
**Pull request by ortem**
Add `RsAssignToImmutableInspection` corresponds to [E0594](https://doc.rust-lang.org/error-index.html#E0594) Rust error.

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/inspections/RsAssignToImmutableInspection.kt [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                }
+        }
+    }
+
+    private fun registerProblem(holder: ProblemsHolder, expr: RsExpr, message: String) {
+        holder.registerProblem(expr, "Cannot assign to $message [E0594]")
+    }
+}
+
+private fun RsExpr?.isAssignBinaryExpr(): Boolean {
```
Comment (fedochet) [inline src/main/kotlin/org/rust/ide/inspections/RsAssignToImmutableInspection.kt]: Why there is a `RsExpr?`, not just `RsExpr`?

File: src/main/kotlin/org/rust/ide/inspections/RsAssignToImmutableInspection.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    val op = this as? RsBinaryExpr ?: return false
```
Comment (fedochet) [inline src/main/kotlin/org/rust/ide/inspections/RsAssignToImmutableInspection.kt]: Maybe `if (this !is RsBinaryExpr) return false`? Don't know what is more readable though

Comment (fedochet) [reply src/main/kotlin/org/rust/ide/inspections/RsAssignToImmutableInspection.kt]: Maybe it is better to return `this is BinaryExpr && this.operatorType is AssignmentOp`?

File: src/main/kotlin/org/rust/ide/inspections/RsAssignToImmutableInspection.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    private fun checkAssignment(expr: RsBinaryExpr, holder: ProblemsHolder) {
+        val left = unwrapParenExprs(expr.left)
+        if (left.isMutable) return
+
+        when (left) {
+            is RsDotExpr -> registerProblem(holder, expr, "field of immutable binding")
+            is RsIndexExpr -> registerProblem(holder, expr, "indexed content of immutable binding")
+            is RsUnaryExpr ->
+                if (left.isDereference) when (left.expr?.type) {
```
Comment (fedochet) [inline src/main/kotlin/org/rust/ide/inspections/RsAssignToImmutableInspection.kt]: This looks odd; consider wrapping this `when` in braces

File: src/main/kotlin/org/rust/lang/core/types/Extensions.kt [hunk 4]
```diff
@@ -80,5 +80,16 @@ fun Ty.builtinDeref(explicit: Boolean = true): Pair<Ty, Mutability>? =
 
 val RsExpr.isDereference: Boolean get() = this is RsUnaryExpr && this.mul != null
 
-
 val RsExpr.isMutable: Boolean get() = mutabilityCategory?.isMutable ?: Mutability.DEFAULT_MUTABILITY.isMut
+
+tailrec fun unwrapParenExprs(expr: RsExpr): RsExpr =
+    when (expr) {
```
Comment (fedochet) [inline src/main/kotlin/org/rust/lang/core/types/Extensions.kt]: I don't like `if expr`s myself, but there are only two options
We should decide something for such cases

File: src/main/kotlin/org/rust/lang/core/types/Extensions.kt [hunk 5]
```diff
@@ -80,5 +80,16 @@ fun Ty.builtinDeref(explicit: Boolean = true): Pair<Ty, Mutability>? =
...
+        is RsParenExpr -> unwrapParenExprs(expr.expr)
+        else -> expr
+    }
+
+val RsExpr.isAssignBinaryExpr: Boolean
+    get() {
+        val op = this as? RsBinaryExpr ?: return false
+        return op.operatorType is AssignmentOp
```
Comment (fedochet) [inline src/main/kotlin/org/rust/lang/core/types/Extensions.kt]: How about `return op is RsBinaryExpr && op.operatorType is AssignmentOp`?

File: src/main/kotlin/org/rust/lang/core/types/Extensions.kt [hunk 6]
```diff
@@ -78,7 +78,12 @@ fun Ty.builtinDeref(explicit: Boolean = true): Pair<Ty, Mutability>? =
...
-
+val RsUnaryExpr.isDereference: Boolean get() = this.mul != null
 
 val RsExpr.isMutable: Boolean get() = mutabilityCategory?.isMutable ?: Mutability.DEFAULT_MUTABILITY.isMut
+
+tailrec fun unwrapParenExprs(expr: RsExpr): RsExpr =
+    if (expr is RsParenExpr) unwrapParenExprs(expr.expr) else expr
+
+val RsExpr.isAssignBinaryExpr: Boolean
+    get() = this is RsBinaryExpr && this.operatorType is AssignmentOp
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/types/Extensions.kt]: Let's move these functions to `RsExpr.kt`.
`org.rust.lang.core.types.Extensions.kt` is for extending PSI with types. All simple PSI extensions should be located inside `org.rust.lang.core.psi.ext` package

**Comment by Undin** [comment]
Interesting
https://doc.rust-lang.org/error-index.html doesn't contain E0594 error although compiler produces it

**Pull request merged**